src/state.rs

```diff
@@ -118,17 +157,24 @@ impl State {
     /// On^2 for n squares
     pub fn gen_moves(&self) -> Vec<Move> {
         let coords = self.board.coords();
-        coords
+        let castles = [true, false].iter().filter_map(|&kingside| {
+            if self.can_castle(kingside) {
```

Comment:
Could `can_castle` and `build_castle_move` be a single fn that returns an optional move?

Reply:
i don't know if this is what you were suggesting, but i do like pulling out `make_castle_move` and `make_simple_move`: 966809ee20021be68c0c14041401fbcae9fffc16 basically still has the same logic, because interestingly those functions don't really share any logic!